parsing/src/recipes.ts

```diff
@@ -142,32 +139,138 @@ function getRecipes(
+                power: Math.round(recipe.mPowerProduction), // generated power - can be rounded to the nearest whole number (all energy numbers are whole numbers) 
+            };   
+            let supplementalRatio: number = Number(recipe.mSupplementalToPowerRatio);
+            // 1. Generator MW generated. This is an hourly value.
+            // 2. Divide by 60, to get the minute value
+            // 3. Now calculate the MJ, using the MJ->MW constant (1/3600), (https://en.wikipedia.org/wiki/Joule#Conversions) 
+            // 4. Now divide this number by the part energy to calculate how many pieces per min
+            const powerMJ: number = (recipe.mPowerProduction / 60) / (1/3600)
+
+            const fuels: Fuel[] = recipe.mFuel       
+            fuels.forEach((fuel: any) => {
+                let fuelItem: Fuel = {
+                    primaryFuel: getPartName(fuel.mFuelClass),
+                    supplementalResource: fuel.mSupplementalResourceClass ? getPartName(fuel.mSupplementalResourceClass) : "",
+                    byProduct: fuel.mByproduct ? getPartName(fuel.mByproduct) : "",
+                    byProductAmount: fuel.mByproductAmount ? Number(fuel.mByproductAmount) : 0
+                }
+
+                //Find the part for the primary fuel
+                let extractedPartText: string = getPartName(fuelItem.primaryFuel);
+                const primaryFuelPart: Part = parts.parts[extractedPartText];
+                let primaryPerMin: number = 0; 
+                if (primaryFuelPart.energyGeneratedInMJ > 0) {
+                    // The rounding here is important to remove floating point errors that appear with some types 
+                    // (this is step 4 from above)
+                    primaryPerMin = parseFloat((powerMJ / primaryFuelPart.energyGeneratedInMJ).toFixed(4))
+                }
+                let primaryAmount : number = 0;
+                if (primaryPerMin > 0) {                        
+                    primaryAmount = primaryPerMin / 60;
+
+                    const ingredients = <any>[];
+                    ingredients.push(
+                        { 
+                            part: fuelItem.primaryFuel,
+                            amount: primaryAmount,
+                            perMin: primaryPerMin
+                        }
+                    )
+                    if (fuelItem.supplementalResource) {
+                        ingredients.push(
+                            { 
+                                part: fuelItem.supplementalResource,
+                                amount: (3 / 50) * supplementalRatio * building.power / 60,
+                                perMin: (3 / 50) * supplementalRatio * building.power // Calculate the ratio of the supplemental resource to the primary fuel
+                            }
+                        )
+                    }
+                    
+                    const products = <any>[];
```

Comment:
Again this should be interfaced, by the looks of it ingredients and products are similar, so you could use an interface where the `isByProduct` is optional, or on products is set to false, as right now it's undefined and may cause errors, and because you are using an `any` type, typescript won't help you if `isByProduct` is called upon and doesn't exist.